examples/stm32/f4/stm32f429i-discovery/lcd-dma/lcd-spi.c

```diff
@@ -0,0 +1,544 @@
+ * byte. Since we're using the SPI port in write_only mode I am not
+ * collecting the "received" bytes into a buffer, but one could of
+ * course. I keep track of how many bytes should have been returned
+ * by decrementing the 'rx_pend' volatile. When it reaches 0 we know
+ * we are done.
+ */
+
+volatile int rx_pend;
+volatile uint16_t spi_rx_buf;
+
+/*
+ * This is the ISR we use. Note that the name is based on the name
+ * in the irq.json file of libopencm3 plus the "_isr" extension.
+ */
+void
+spi5_isr(void) {
+    spi_rx_buf = SPI_DR(SPI5);
+    --rx_pend;
+}
+
+/* Simple double buffering, one frame is displayed, the
+ * other being built.
+ */
+uint16_t *cur_frame;
+uint16_t *display_frame;
+
+
+/*
+ * Drawing a pixel consists of storing a 16 bit value in the
+ * memory used to hold the frame. This code computes the address
+ * of the word to store, and puts in the value we pass to it.
+ */
+void
+lcd_draw_pixel(int x, int y, uint16_t color) {
+    *(cur_frame + x + y * LCD_WIDTH) = color;
+}
+
+/*
+ * Fun fact, same SPI port as the MEMS example but different
+ * I/O pins. Clearly you can't use both the SPI port and the
+ * MEMS chip at the same time in this example.
+ *
+ * For the STM32-DISCO board, SPI pins in use:
+ *  N/C - RESET
+ *  PC2 - CS (could be NSS but won't be)
+ *  PF7 - SCLK (AF5) SPI5
+ *  PD13 - DATA / CMD*
+ *  PF9 - MOSI (AF5) SPI5
+ */
+
+/*
+ * This structure defines the sequence of commands to send
+ * to the Display in order to initialize it. The AdaFruit
+ * folks do something similar, it helps when debugging the
+ * initialization sequence for the display.
+ */
+struct tft_command {
+	uint16_t delay;		// If you need a delay after
+	uint8_t cmd;		// command to send
+	uint8_t n_args;		// How many arguments it has
+};
+
+
+/* prototype for lcd_command */
+static void lcd_command(uint8_t cmd, int delay, int n_args,
+						const uint8_t *args);
+
+/*
+ * void lcd_command(cmd, delay, args, arg_ptr)
+ *
+ * All singing all dancing 'do a command' feature. Basically it
+ * sends a command, and if args are present it sets 'data' and
+ * sends those along too.
+ */
+static void
+lcd_command(uint8_t cmd, int delay, int n_args, const uint8_t *args) {
+	uint32_t timeout;
+	int i;
+
+	gpio_clear(GPIOC, GPIO2);	// Select the LCD
+	rx_pend++;
+	spi_send(SPI5, cmd);
+	/* We need to wait until it is sent, if we turn on the Data
+	 * line too soon, it ends up confusing the display to thinking
+	 * its a data transfer, as it samples the D/CX line on the last
+	 * bit sent.
+	 */
+	for (timeout = 0; (timeout < 1000) && (rx_pend); timeout++);
+	rx_pend = 0; // sometimes, at 10Mhz we miss this
+	if (n_args) {
+		gpio_set(GPIOD, GPIO13);	// Set the D/CX pin
+		for (i = 0; i < n_args; i++) {
+			rx_pend++;
+			spi_send(SPI5, *(args+i));
+		}
+		/* This wait so that we don't pull CS too soon after
+		 * sending the last byte of data.
+		 */
+		for (timeout = 0; (timeout < 1000) && (rx_pend); timeout++);
```

Comment:
No, TXE is set as soon as transfer starts, since the data register is at that point ready to accept another word. The right way to wait for transfer to end is to wait for TXE=1 and BSY=0.